packages/record-data/addon/-private/relationships/state/has-many.ts

```diff
@@ -156,13 +156,45 @@ export default class ManyRelationship extends Relationship {
     this.notifyHasManyChange();
   }
 
-  computeChanges(recordDatas: RelationshipRecordData[] = []) {
-    const members = this.canonicalMembers.toArray();
-    for (let i = members.length - 1; i >= 0; i--) {
-      this.removeCanonicalRecordData(members[i], i);
+  computeChanges(newMembers: RelationshipRecordData[] = []) {
+    const oldMembers = this.canonicalMembers.toArray();
+
+    let oldLength = oldMembers.length;
+    let newLength = newMembers.length;
+    let shortest = Math.min(oldLength, newLength);
+    // find the difference block in the old data
+    // from the first change to the last change
+    let blockStart = shortest;
+    // find the first difference
+    for (let i = 0; i < shortest; i++) {
+      if (oldMembers[i] !== newMembers[i]) {
+        blockStart = i;
+        break;
+      }
+    }
+    let oldBlockLength = oldLength - blockStart;
+    let newBlockLength = newLength - blockStart;
+    if (blockStart < shortest) {
+      // we found a difference so
+      // find the last difference
+      for (let i = oldLength - 1, j = newLength - 1; i >= oldLength - shortest; i--, j--) {
+        if (oldMembers[i] !== newMembers[j]) {
+          oldBlockLength = i - blockStart + 1;
+          newBlockLength = j - blockStart + 1;
+          break;
+        }
+      }
+    }
+
+    // remove old records from the change block
+    if (oldBlockLength > 0) {
+      for (let i = blockStart + oldBlockLength - 1; i >= blockStart; i--) {
+        this.removeCanonicalRecordData(oldMembers[i], i, true);
+      }
     }
-    for (let i = 0, l = recordDatas.length; i < l; i++) {
-      this.addCanonicalRecordData(recordDatas[i], i);
+    // insert new records into the change block
+    for (let i = blockStart; i < blockStart + newBlockLength; i++) {
+      this.addCanonicalRecordData(newMembers[i], i);
```

Comment:
This is not about Set behaviour, but about Array behaviour. This starts by finding what's changed (add, remove, move), then within that block, blast it and rewrite it. The block referred to is from the first change to the last change.

There are more performance improvements I can see, but leaving those for a second PR to sit on top of this one.

Reply:
From my perusing, `canonicalMembers` is a Set (`OrderedSet`) btw.  We short circuit `addCanonicalRecordData` if `canonicalMembers` already `has` the specific recordData in the set.  

e.g. if you left the second for loop as is, I imagine your tests still pass.  I'm just making a note b/c it could help us think about an alternative solution as well.

Reply:
@snewcomer we don't use those ordered sets effectively at all.  This bug has been an issue I've wanted (and tried) to clean up as part of refactors many times but keeps getting lost or going stale, it's actually why I added TODO tests that would let us know if it was fixed!

Reply:
> This bug
The one I'm addressing in this PR?
What can I do to help land this?

Reply:
> not considering performance

I've considered it a lot in writing this